File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, sub) {
        if (!sub) {
            return true;
        } 

        if (!root) {
            return false;
        }

        const same = this.isSame(root, sub);
        if (same) {
            return true;
        }

        const left = this.isSubtree(root.left, sub);
        const right = this.isSubtree(root.right, sub);

        return left || right;
    }

    isSame(root, sub) {
        if (!root && !sub) {
            return true;
        }

        if (root && sub && root.val === sub.val) {
            const left = this.isSame(root.left, sub.left);
            const right = this.isSame(root.right, sub.right);

            return left && right;
        }

        return false;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, prev_max):
            if not node:
                return 0    

            is_good = node.val >= prev_max
            new_max = max(prev_max, node.val)

            left_count = dfs(node.left, new_max)
            right_count = dfs(node.right, new_max)

            return is_good + left_count + right_count
        
        return dfs(root, root.val)